# Thông tin nhóm 14

1612406 - Đặng Phương Nam

1612423 - Lê Minh nghĩa

# 1. Câu hỏi

Cho các thông tin về căn nhà:

- Giá cho thuê trước khi bán.
- Địa chỉ.
- Loại nhà.
- Số phòng ngủ.
- Số nhà vệ sinh.
- Diện tích phần nhà.
- Diện tích toàn bộ lô đất.
- Năm xây dựng.
- Tiền thuế.
- Gần mấy trường học.
- Tình hình tội phạm.
- ...

Hỏi giá trị của căn nhà là bao nhiêu tiền?

# 2. Lợi ích khi trả lời được câu hỏi

Nhờ vào thông tin của căn nhà:

- Người bán có thể dự đoán được giá trị căn nhà mà mình muốn bán.
- Người mua có thể ước lượng được căn nhà mình muốn mua có giá cả hợp lý hay không?.
- Dự đoán được giá trị căn nhà của mình.
- ...

# 3. Thu thập dữ liệu

## 3.1. Parse HTML

Dữ liệu thu thập từ trang web https://www.realtytrac.com/. Ta chỉ thu thập dữ liệu "các estate đã được bán tại  California", ta có HTML cần parse: https://m.realtytrac.com/mapsearch/sold/ca/

Thời gian lấy dữ liệu: Ngày 10/12/2019

In [ ]:
# Import các thư viện cần thiết.
import urllib.robotparser
import json
import os.path
import re

from requests_html import HTMLSession

In [ ]:
rp = urllib.robotparser.RobotFileParser()
rp.set_url('https://m.realtytrac.com/robots.txt')
rp.read()

In [ ]:
# Lấy link chứa danh sách các ngôi nhà đã bán theo Quận tại California
check_parse = rp.can_fetch('*', 'https://m.realtytrac.com/mapsearch/sold/ca/')

base_url = "https://m.realtytrac.com"
list_url_county = []

# Kiểm tra việc parse HTML được cho phép hay không?
if check_parse == True:
    session = HTMLSession()
    r = session.get('https://m.realtytrac.com/mapsearch/sold/ca/')

    counties = r.html.find("option")

    for county in counties:
        tail_url = county.attrs['value']
        if (tail_url):
            url = base_url + tail_url
            list_url_county.append(url)


list_url_county

In [ ]:
# Hàm trả về generator, generator này trả về trang tiếp theo sau mỗi lần duyệt.
def urlGenerator(baseUrl, startPage=1):
    i = startPage
    if i <= 1:
        yield baseUrl
        i = 2
    while True:
        yield f"{baseUrl}/p-{i}"
        i += 1


# Hàm parseDetailPage: parse để lấy thông tin chi tiết của từng căn nhà từ detailUrl (của căn nhà đã bán).
# Tham số: session và detailUrl của căn nhà cần parse.
# Trả về: dictionary chứa các thông tin đã parse được.
def parseDetailPage(session: HTMLSession, detailUrl: str):

    # Lấy mã HTML của trang web.
    r = session.get(detailUrl)

    # Dictionary lưu lại kết quả.
    result = {}

    # Lấy address của estate.
    address = {}
    street = r.html.find("[itemprop = 'streetAddress']", first=True)
    if street is not None:
        street = street.text

    locality = r.html.find("[itemprop = 'addressLocality']", first=True)
    if locality is not None:
        locality = locality.text

    region = r.html.find("[itemprop = 'addressRegion']", first=True)
    if region is not None:
        region = region.text

    code = r.html.find("[itemprop = 'postalCode']", first=True)
    if code is not None:
        code = code.text

    address["street"] = street
    address["locality"] = locality
    address["region"] = region
    address["code"] = code

    result["address"] = address

    # Lấy thông tin ngày đã bán căn nhà.
    date_sold = r.html.find(".recently-sold", first=True)
    if date_sold is not None:
        result["date_sold"] = date_sold.text.replace("SOLD ON ", "")
    else:
        result["date_sold"] = None

    # Lấy giá cho thuê trước khi bán.
    mortgage = r.html.find(".property-price-info", first=True)
    if mortgage is not None:
        result["mortgage"] = mortgage.text.replace("Est. Mortgage: ","")
    else:
        result["mortgage"] = None


    # Lấy phân mô tả chi tiết về căn nhà.
    details = []
    detailTexts = r.html.find(".detail-text")
    if detailTexts is not None:
        for detailText in detailTexts:
            details.append(detailText.html)  # sở dĩ lấy mà HTML để dễ parse sau này (khi muốn lấy thông tin từ details)
    result["details"] = details

    # Lấy property info của căn nhà.
    propertyInfo = r.html.find(".property-info", first=True)
    if propertyInfo is not None:
        info = {}
        items = propertyInfo.find(".item")
        if items is not None:
            for item in items:
                name = item.find(".name", first=True).text.lower().replace(" ", "_")
                value = item.find(".value", first=True).text
                info[name] = value

        result["info"] = info
    else:
        result["info"] = None

    # Lấy property taxes của căn nhà.
    property_taxes = r.html.find(".property-taxes")
    if property_taxes is not None:
        for tax in property_taxes:
            key = tax.find(".section-head", first=True).text.lower().replace(" ", "_")
            taxes = {}
            items = tax.find(".item")
            if items is not None:
                for item in items:
                    name = item.find(".name", first=True).text.lower().replace(" ", "_")
                    value = item.find(".value", first=True).text
                    taxes[name] = value

            result[key] = taxes

    # Lấy số lượng trường học gần đó.
    local_school = r.html.find(".property-schools", first=True)
    num_school = 0
    if local_school is not None:
        num_school = len(local_school.find(".schoolInfo"))

    result["school"] = num_school

    # Lấy thông tin tội phạm trong vùng.
    crimes = {}
    local_crime_index = r.html.find(".property-local-crime", first=True)
    if local_crime_index is not None:
        type_crimes = local_crime_index.find(".gradeTitle")
        if type_crimes is not None:
            for type_crime in type_crimes:
                match = re.match(r"(.+?) = (\d+)", type_crime.text)
                if match is not None:
                    key = match.group(1).lower().replace(' ', '_')
                    value = int(match.group(2))
                    crimes[key] = value

    result["local_crime_index"] = crimes

    # Lấy số lượng nhà bị tịch thu gần đó.
    nearby_foreclosures = r.html.find(".property-nearby-fc", first=True)
    num_fc = 0
    if nearby_foreclosures is not None:
        num_fc = len(nearby_foreclosures.find(".nearby-property"))

    result["foreclosures"] = num_fc

    return result

In [ ]:
# Hàm parseSearchPage: parse theo page (mỗi parse chứa danh sách và url của 25 căn nhà đã được bán).
# Tham số: session và url của page hiện tại.
# Trả về: list thông tin chi tiết của tất cả căn nhà tại page hiện tại và url của page kế tiếp.
def parseSearchPage(session: HTMLSession, urlPage: str):

    # Lấy mã HTML.
    r = session.get(urlPage)

    # Lấu detailUrl của tất cả bất động sản.
    detailUrls = [link for link in r.html.absolute_links if 'property' in link]

    results = []  # List chứa kết quả parse được tại page.

    # Tiến hành parse từ detailUrl
    for detailUrl in detailUrls:
        # Kiểm tra việc parse HTML được cho phép hay không?
        if not rp.can_fetch('*', detailUrl):
            print(f"SKIP: {detailUrl}")
            continue

        print(f"Parsing detail url {detailUrl}")
        result = parseDetailPage(session, detailUrl)
        results.append(result)

    nextPageAnchor = r.html.find(".current + .page", first=True)
    return results, nextPageAnchor is not None

In [ ]:
# Hàm getAllCountyProperties: parse tất cả các page của một County để lấy thông tin chi tiết của 
# căn nhà được bán thành công.
# Tham số: session và countryUrl chứa thông tin các căn nhà được bán của một County.
# Trả về: File chứa kết quả đã parse thành công (tên file là tên County được parse).
def getAllCountyProperties(session: HTMLSession, countyUrl: str):
    # Lấy tên County từ urlCounty.
    if countyUrl.endswith("/"):
        countyUrl = countyUrl[:-1]
    fileName = f"data/{os.path.basename(countyUrl)}.json"

    # Kiểm tra sự tồn tại file.
    if os.path.exists(fileName):
        print(f"FILE EXISTS: {fileName}")
        return

    # Mở file để ghi dữ liệu (kiểu file là json)
    with open(fileName, "a") as fOut:
        for url in urlGenerator(countyUrl, 1):
            # Kiểm tra việc parse HTML được cho phép hay không?
            if not rp.can_fetch('*', url):
                print(f"SKIP: {url}")
                continue

            # Tiến hành parse theo từng page.
            print(f"Parsing {url}")
            results, nextPage = parseSearchPage(session, url)

            # Ghi kết quả xuống file.
            for result in results:
                json.dump(result, fOut, ensure_ascii=False)
                fOut.write("\n")

            # Kiểm tra có còn page tiếp theo hay không?
            if not nextPage:
                break

            # Mỗi lần parse 1 page thì cho sleep 2s.
            time.sleep(2)

Tiến hành lấy dữ liệu. Lưu ý trong quá trình lấy có thể bị timeout do đường truyền mạng, nếu bị thì chỉ cần làm theo các bước sau:
- Xem đã lấy tới page mấy. vd: "...page-125"
- Đổi lại tên file trong thư mục data là "orange-county-1.json" chẳng hạn (số 1, 2, 3,.. là theo mỗi lần lấy dữ liệu của mình).
- Vào hàm "getAllCountyProperties" đổi lại dòng "for url in urlGenerator(countyUrl, 1)" thành "for url in urlGenerator(countyUrl, 125)", số 125 là tùy mình lấy tới page thứ mấy mà bị timeout.
- Dữ liệu tất nhiên sẽ bi trùng, cứ việc yên tâm vì phần sau sẽ xóa trung.

In [ ]:
url = "https://m.realtytrac.com/mapsearch/sold/ca/orange-county"
session = HTMLSession()
getAllCountyProperties(session, url)

In [ ]:
# Phần gộp từng phần dữ liệu đã lấy dang dở lại thành 1 file.
# Nếu không bị timeout như trường hợp ở trên thì không dùng shell code này.
inputFiles = ["data/orange-county-1.json", "data/orange-county-2.json", 
              "data/orange-county-3.json", "data/orange-county-4.json"]
outputFile = "data/full-orange-county.json"

with open(outputFile, "a+") as fOut:
    for inputFile in inputFiles:
        with open(inputFile, "r") as fIn:
            data = fIn.read()
            fOut.write(data)

In [ ]:
# Phần convert file json sang file csv.

def xstr(s):
    return '' if s is None else str(s)


# Hàm preprocessFile: chuyển file json sang file csv.
# Tham số: tên file json và tên file csv.
# Không trả về.
def preprocessFile(inputFile: str, outputFile: str):

    # Kiểm tra tồn tại file Input.
    if os.path.exists(outputFile):
        print(f"FILE EXISTS: {outputFile}")
        return

    # Kiểm tra tồn tại file Output.
    if not os.path.exists(inputFile):
        print(f"Input file not found: {inputFile}")
        return

    with open(inputFile) as fIn, open(outputFile, "w") as fOut:
        tab = "\t"
        headers = (
            "address_street",
            "address_locality",
            "address_region",
            "address_code",
            "date_sold",
            "mortgage",
            "info_type",
            "info_bedrooms",
            "info_bathrooms",
            "info_size",
            "info_lot_size",
            "info_year_built",
            "info_est_value",
            "info_sold_price",
            "info_property_id",
            "info_county",
            "info_parcel_number",
            "taxes_land",
            "taxes_improvements",
            "taxes_total",
            "taxes_taxes",
            "school",
            "total_crime",
            "violent_crime",
            "property_crime",
            "foreclosures",
        )

        # Ghi các tên cột vào file output, mỗi cột ngăn cách bởi "\t".
        fOut.write(f"{tab.join(headers)}\n")

        # Duyệt từng line trong file Input (json).
        for line in fIn:
            row = json.loads(line)

            # Lấy địa chỉ
            address = row.get("address")
            address_street = address.get("street")
            address_locality = address.get("locality")
            address_region = address.get("region")
            address_code = address.get("code")

            # Lấy ngày bán thành công.
            date_sold = row.get("date_sold")

            # Lấy giá cho thuê mỗi tháng (trước khi được bán).
            mortgage = row.get("mortgage")
            if mortgage is not None:
                mortgage = float(mortgage.replace("$","").replace(",","").replace("/mo","").replace("Est. Refinance: ", ""))

            # Lấy phần info
            info = row.get("info")

            # Type.
            info_type = info.get("type")

            # Bedrooms.
            info_bedrooms = info.get("bedrooms")
            if info_bedrooms == "Contact Agent":
                info_bedrooms = None
            else:
                info_bedrooms = float(info_bedrooms)

            # Bathrooms.
            info_bathrooms = info.get("bathrooms")
            if info_bathrooms == "Contact Agent":
                info_bathrooms = None
            else:
                info_bathrooms = float(info_bathrooms)

            # Size.
            info_size = info.get("size")
            if info_size == "Contact Agent":
                info_size = None
            else:
                info_size = float(info_size.replace(",", "").replace(" sqft", ""))

            # Lot size.
            info_lot_size = info.get("lot_size")
            if info_lot_size == "Contact Agent":
                info_lot_size = None
            else:
                info_lot_size = float(info_lot_size.replace(",", "").replace(" sqft", "").replace(" acres", ""))

            # Year build.
            info_year_built = info.get("year_built")
            if info_year_built  == "Contact Agent":
                info_year_built = None
            else:
                info_year_built = int(info_year_built)

            # Est value.
            info_est_value = info.get("est._value")
            if info_est_value is not None:
                info_est_value = float(info_est_value.replace("$", "").replace(",",""))

            # Sold price.
            info_sold_price = info.get("sold_price")
            if info_sold_price is not None:
                if info_sold_price != "N/A":
                    info_sold_price = float(info_sold_price.replace("$", "").replace(",", ""))

            # ID.
            info_property_id = info.get("property_id")

            # County.
            info_county = info.get("county")

            # Parcel_number.
            info_parcel_number = info.get("parcel_number")

            # Lấy phần Taxes.
            taxes = row.get("property_taxes")
            taxes_land = None
            taxes_improvements = None
            taxes_total = None
            taxes_taxes = None
            if taxes is not None:
                taxes_land = float(taxes.get("land").replace("$", "").replace(",", ""))
                taxes_improvements = float(taxes.get("improvements").replace("$", "").replace(",", ""))
                taxes_total = float(taxes.get("total").replace("$", "").replace(",", ""))
                taxes_taxes = taxes.get("taxes").replace("$", "").replace(",", "")

            # Lấy số lượng scholl gần đó.
            school = int(row.get("school"))

            # Lấy tình hình tội phạm.
            crime = row.get("local_crime_index")
            if crime is not None:
                total_crime = crime.get("total_crime")
                if total_crime is not None:
                    total_crime = int(total_crime)
                violent_crime = crime.get("violent_crime")
                if violent_crime is not None:
                    violent_crime = int(violent_crime)
                property_crime = crime.get("property_crime")
                if property_crime is not None:
                    property_crime = int(property_crime)

            # Lấy số lượng foreclosures gần đó.
            foreclosures = int(row.get("foreclosures"))

            # Row này là tổng hợp các thông tin lấy được ở trên.
            row = (
                xstr(address_street),
                xstr(address_locality),
                xstr(address_region),
                xstr(address_code),
                xstr(date_sold),
                xstr(mortgage),
                xstr(info_type),
                xstr(info_bedrooms),
                xstr(info_bathrooms),
                xstr(info_size),
                xstr(info_lot_size),
                xstr(info_year_built),
                xstr(info_est_value),
                xstr(info_sold_price),
                xstr(info_property_id),
                xstr(info_county),
                xstr(info_parcel_number),
                xstr(taxes_land),
                xstr(taxes_improvements),
                xstr(taxes_total),
                xstr(taxes_taxes),
                xstr(school),
                xstr(total_crime),
                xstr(violent_crime),
                xstr(property_crime),
                xstr(foreclosures),
            )

            # Ghi xuống file output
            fOut.write(f"{tab.join(row)}\n")

&#9889; Sở dĩ nhóm không thêm phần `description` vào dữ liệu là bởi vì khi quan sát phần mã `html` của `description` thì phát hiện ra nó được viết theo một `format` nhất định, các thông tin số liệu của nó được lấy từ các bảng số liệu trong cùng một trang web (tức là các phần thông tin nhà, thuế,.. mà mình đã parse ở trên). Nếu số liệu nào có thì sẽ được lấy ra và tạo một câu mô tả theo `format` cho trước, còn số liệu nào không có thì hiển nhiên không xuất hiện câu đó. Như vậy, có thể kết luận là phần `description` lúc này sẽ không giúp ích gì cho chúng ta.

In [ ]:
inputFile = "data/full-orange-county.json"
outputFile = "data/orange-county.csv"
preprocessFile(inputFile, outputFile)

## 3.2. Lựa chọn dữ liệu của năm 2019 và có "correct ouput"

In [ ]:
# Import các thư viện cần thiết.
import pandas as pd

### 3.2.1 Xóa dữ liệu trùng

#### Dữ liệu ban đầu

In [ ]:
# Đọc dữ liệu lên
file_csv = "data/Orange-County/orange-county.csv"
data_df =  pd.read_csv(file_csv, sep='\t')
data_df.head(10)

In [ ]:
data_df.info()

Dữ liệu ban đầu có 21268 dòng.

#### Dữ liệu sau khi xóa trùng

In [ ]:
# Xóa các phần dữ liệu bị trùng
unique_data_df = data_df.drop_duplicates()

In [ ]:
# Xem phần dữ liệu khi đã xóa trùng
unique_data_df.info()

Dữ liệu sau khi xóa trùng còn lại 18893 dòng.

In [ ]:
# Ghi dữ liệu xuống file csv
unique_data_df.to_csv("data/Orange-County/delete-duplicate-orange.csv", sep='\t', index=False)

### 3.2.2. Lấy dữ liệu của năm 2019

In [ ]:
# Đọc dữ liệu lên
df = pd.read_csv("data/Orange-County/delete-duplicate-orange.csv", sep='\t')
df.head(10)

In [ ]:
df.info()

In [ ]:
# Thêm cột năm bán vào dữ liệu
df = df.assign(year_sold=df.date_sold.str[-4:])

In [ ]:
# Xem dữ liệu này thu thập của các năm nào
df.year_sold.unique()

In [ ]:
# Chỉ lấy phần dữ liệu của năm 2019
df = df[df.year_sold == "2019"]
df.year_sold = df.year_sold.astype(int)

In [ ]:
# Ghi dữ liệu xuống file csv
df.to_csv("data/Orange-County/delete-duplicate-orange-2019.csv", sep='\t', index=False)

### 3.2.3. Lấy dữ liệu của năm 2019 có "correct output"

In [ ]:
import pandas as pd

In [ ]:
# Đọc dữ liệu
data_df = pd.read_csv("data/Orange-County/delete-duplicate-orange-2019.csv", sep='\t')
data_df.head(10)

In [ ]:
data_df.info()

Dữ liệu có 16011 căn nhà được bán trong 2019

Ta chỉ lấy dữ liệu có `info_sold_price` khác `null`. Như đã nói ở phần cuối của `1.1. Parse HTML` thì không thể tìm kiếm các giá trị chính xác từ phần `decription` để có thể lắp đầy giá trị `null`, mà nếu có tìm cũng chỉ toàn ra giá trị trung bình của các căn nhà được bán lân cận chứ không phải giá trị chính xác của nó.

In [ ]:
df = data_df[pd.notna(data_df["info_sold_price"])]

In [ ]:
df.info()

Dữ liệu còn lại là 15755.

&#9889; Sau khi xóa các dòng có sold_price là `null`. Xem lại cột `info_est_value` thấy toàn `null`, cột này ý nghĩa là giá trị định giá của căn nhà trên các bản rao bán (chưa chốt giá). Nên xóa đi cột này là hoàn toàn hợp lý.

In [ ]:
df = df.drop(columns=["info_est_value"])
df.info()

Ta cần kiểm tra xem ứng với `15755` dòng dữ liệu thì phải có đúng `15755` `info_property_id` hay không?

In [ ]:
len(df.info_property_id.unique())

OK ổn vì có đúng `15755` `info_property_id` như mong muốn

Bây giờ, ta kiểm tra liệu ứng với mỗi `address_street` liệu chỉ có duy nhất một căn nhà được bán thành công hay không?

In [ ]:
address_street_count = df.address_street.value_counts()
address_street_count

Như vậy có `38` `address_street` có số lượng nhà bán thành công trong năm 2019 lơn hơn 1, trong đó lại có `1600 S Disneyland Dr` có số lần xuất hiện lên đến con sô `16`, điều này hơi kỳ lạ. Vì vậy, ta cần quan sát thêm về các `addresss_street` kiểu này.

In [ ]:
# Lấy tên các addresss_street có số lượng nhà được bán trong năm 2019 nhiều hơn 1
duplicate_addresss_street = address_street_count[address_street_count > 1].index
duplicate_addresss_street

In [ ]:
len(duplicate_addresss_street)

Xem thử tất cả các dòng có `address_street` là `1600 S Disneyland Dr` (hay `duplicate_addresss_street[0]`)

In [ ]:
tmp_df = df[df.address_street == duplicate_addresss_street[0]]
tmp_df

In [ ]:
tmp_df.info_sold_price

Đối với những dòng có cùng `address_street` mà lại khác `date_sold` và `info_sold_price` thì nên bỏ là hợp lý và khó có thể lấy được chính xác giá trị `sold_price`.

Tiếp tục, xem thử tất cả các dòng có `address_street` là `200 W Midway Dr ` (hay `duplicate_addresss_street[1]`)

In [ ]:
tmp_df = df[df.address_street == duplicate_addresss_street[1]]
tmp_df

In [ ]:
tmp_df.info_sold_price

Mặc dù có cùng `address_street`, cùng `date_sold` và cùng `info_sold_price` nhưng lại có các giá trị `taxes` thay đổi một cách kỳ lại + với các thông tin chính (`info_bedrooms`, `info_bathroooms`,...) bị thiếu. Phần các thông tin chính này có thể giải quyết bằng `fill miss value` (`mean`, `median`, `most`,...) nhưng phần `taxes` thì nhóm không biết xác định như thế nào cho hợp lý và gần chính xác nhất. Nên quyết định bỏ luôn các dòng dạng này.

In [ ]:
sum(address_street_count[address_street_count > 1])

Có `106` dòng dữ liệu bị như thế, so với `15755` dòng dữ liệu hiện có thì chúng chưa chiếm đến `1%` nên nhóm quyết định bỏ đi là hợp lý.

In [ ]:
df = df[~df.address_street.isin(duplicate_addresss_street)]
df.info()

In [ ]:
# Ghi dữ liệu xuống file csv.
df.to_csv("data/Orange-County/data-2019.csv", sep='\t', index=False)

#### Tách dữ liệu thành 2 tập train và test theo tỉ lệ 80% 20%

In [ ]:
# Đọc dữ liệu lên
data_df = pd.read_csv("data/Orange-County/data-2019.csv", sep='\t')
data_df.head()

In [ ]:
data_df.info()

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
# Đổi tên
data_df = data_df.rename(columns={"info_sold_price": "sold_price"})

In [ ]:
# Tách dữ liệu theo tỉ lệ 80% và 20%
train, test = train_test_split(data_df, test_size=0.2, random_state=0)
train.head().index

&#9889; Nhóm đã cố định `random_state` trong `train_test_split` để đảm bảo kết quả của thầy ra giống với của nhóm. Kết quả của câu lệnh `train.head().index` của nhóm ra 5 giá trị là: [2746, 8859, 11288, 2153, 12119]. 

In [ ]:
train.info()

In [ ]:
test.info()

In [ ]:
# Ghi tập train
train.to_csv("data/Orange-County/train-data-2019.csv", sep='\t', index=False)

In [ ]:
# Ghi tập test
test.to_csv("data/Orange-County/test-data-2019.csv", sep='\t', index=False)

# 4. Tiền xử lý dữ liệu

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

In [2]:
data_df = pd.read_csv("data/Orange-County/train-data-2019.csv", sep='\t')
data_df.head()

address_street  address_locality address_region  address_code  \
0        13242 Amarillo Dr       Westminster             CA         92683   
1        716 S Palomino Ln           Anaheim             CA         92807   
2      19702 Lancewood Plz       Yorba Linda             CA         92886   
3  25885 Trabuco Rd Apt 59       Lake Forest             CA         92630   
4       6742 Gate Hill Cir  Huntington Beach             CA         92648   

    date_sold  mortgage                info_type  info_bedrooms  \
0  10/24/2019    3157.0  Single Family Residence            3.0   
1  08/16/2019    2748.0  Single Family Residence            3.0   
2  07/23/2019    2966.0  Single Family Residence            4.0   
3  10/30/2019    1726.0              Condominium            2.0   
4  07/15/2019    5860.0  Single Family Residence            4.0   

   info_bathrooms  info_size  ...  taxes_land  taxes_improvements  \
0             1.0     1090.0  ...    508260.0             54780.0   
1             3.0     1582.0  ...    371674.0            148526.0   
2             2.0     1440.0  ...     29459.0             43373.0   
3             2.0     1190.0  ...    112879.0             93656.0   
4             3.0     2193.0  ...    159764.0            148957.0   

   taxes_total    taxes_taxes school  total_crime  violent_crime  \
0     563040.0  6969 (1.23 %)      3         97.0           79.0   
1     520200.0  5527 (1.06 %)      3         88.0           71.0   
2      72832.0  1240 (1.70 %)      3         22.0           12.0   
3     206535.0  2096 (1.01 %)      3         29.0           23.0   
4     308721.0  3706 (1.20 %)      3         57.0           47.0   

   property_crime  foreclosures year_sold  
0           140.0             4      2019  
1           127.0             4      2019  
2            44.0             4      2019  
3            42.0             4      2019  
4            82.0             4      2019  

[5 rows x 26 columns]

In [3]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12519 entries, 0 to 12518
Data columns (total 26 columns):
address_street        12485 non-null object
address_locality      12519 non-null object
address_region        12519 non-null object
address_code          12519 non-null int64
date_sold             12519 non-null object
mortgage              12519 non-null float64
info_type             12519 non-null object
info_bedrooms         11363 non-null float64
info_bathrooms        11371 non-null float64
info_size             11946 non-null float64
info_lot_size         8484 non-null float64
info_year_built       11571 non-null float64
sold_price            12519 non-null float64
info_property_id      12519 non-null int64
info_county           12519 non-null object
info_parcel_number    12519 non-null int64
taxes_land            12321 non-null float64
taxes_improvements    12321 non-null float64
taxes_total           12321 non-null float64
taxes_taxes           12321 non-null object
schoo

Tập train có `12519` dòng dữ liệu và `26` cột.

In [4]:
data_df.describe(include="all")

address_street address_locality address_region  address_code  \
count                12485            12519          12519  12519.000000   
unique               12485               52              2           NaN   
top     2127 W Dogwood Ave           Irvine             CA           NaN   
freq                     1             1107          12517           NaN   
mean                   NaN              NaN            NaN  92552.015816   
std                    NaN              NaN            NaN    564.557227   
min                    NaN              NaN            NaN  90620.000000   
25%                    NaN              NaN            NaN  92630.000000   
50%                    NaN              NaN            NaN  92677.000000   
75%                    NaN              NaN            NaN  92804.000000   
max                    NaN              NaN            NaN  92887.000000   

         date_sold      mortgage                info_type  info_bedrooms  \
count        12519  1.251900e+04                    12519   11363.000000   
unique         155           NaN                       14            NaN   
top     06/28/2019           NaN  Single Family Residence            NaN   
freq           232           NaN                     8162            NaN   
mean           NaN  8.928697e+03                      NaN       3.124263   
std            NaN  7.526277e+04                      NaN       1.008088   
min            NaN  2.000000e+00                      NaN       1.000000   
25%            NaN  2.548000e+03                      NaN       2.000000   
50%            NaN  3.339000e+03                      NaN       3.000000   
75%            NaN  4.610000e+03                      NaN       4.000000   
max            NaN  2.078417e+06                      NaN       8.000000   

        info_bathrooms      info_size  ...    taxes_land  taxes_improvements  \
count     11371.000000   11946.000000  ...  1.232100e+04        1.232100e+04   
unique             NaN            NaN  ...           NaN                 NaN   
top                NaN            NaN  ...           NaN                 NaN   
freq               NaN            NaN  ...           NaN                 NaN   
mean          2.602102    2426.583208  ...  5.009780e+05        2.826592e+05   
std           0.991297    7797.643516  ...  1.434805e+06        1.751883e+06   
min           1.000000     301.000000  ...  4.410000e+02        0.000000e+00   
25%           2.000000    1264.000000  ...  1.504430e+05        7.841300e+04   
50%           3.000000    1671.000000  ...  3.111640e+05        1.312930e+05   
75%           3.000000    2377.750000  ...  5.119080e+05        2.319820e+05   
max          11.000000  450854.000000  ...  6.929106e+07        9.338348e+07   

         taxes_total     taxes_taxes        school   total_crime  \
count   1.232100e+04           12321  12519.000000  12156.000000   
unique           NaN           11959           NaN           NaN   
top              NaN  10555 (1.12 %)           NaN           NaN   
freq             NaN              19           NaN           NaN   
mean    7.836531e+05             NaN      2.884016     54.516946   
std     2.832966e+06             NaN      0.578382     29.778698   
min     0.000000e+00             NaN      0.000000     13.000000   
25%     2.769060e+05             NaN      3.000000     29.000000   
50%     4.712400e+05             NaN      3.000000     55.000000   
75%     7.320000e+05             NaN      3.000000     71.000000   
max     1.447999e+08             NaN      3.000000    128.000000   

        violent_crime  property_crime  foreclosures year_sold  
count    12156.000000    12156.000000  12519.000000   12519.0  
unique            NaN             NaN           NaN       NaN  
top               NaN             NaN           NaN       NaN  
freq              NaN             NaN           NaN       NaN  
mean        42.570171       82.203274      3.853822    2019.0  
std         28.960716   

## 4.1. Tách tập dữ liệu thành 2 phần train (70%) và validation (30%)

In [5]:
# Tách X và y
y_sr = data_df["sold_price"] # sr là viết tắt của series
X_df = data_df.drop("sold_price", axis=1)

In [6]:
# Tách tập train và tập validation theo tỉ lệ 70%:30%
train_X_df, val_X_df, train_y_sr, val_y_sr = train_test_split(X_df, y_sr, test_size=0.3, random_state=0)
train_X_df.head().index

Int64Index([10747, 12214, 5916, 489, 12257], dtype='int64')

&#9889; Nhóm đã cố định `random_state` trong `train_test_split` để đảm bảo kết quả của thầy ra giống với của nhóm. Kết quả của câu lệnh `train_X_df.head().index` của nhóm ra 5 giá trị là: [10747, 12214, 5916, 489, 12257]. 

In [7]:
train_X_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8763 entries, 10747 to 2732
Data columns (total 25 columns):
address_street        8736 non-null object
address_locality      8763 non-null object
address_region        8763 non-null object
address_code          8763 non-null int64
date_sold             8763 non-null object
mortgage              8763 non-null float64
info_type             8763 non-null object
info_bedrooms         7951 non-null float64
info_bathrooms        7956 non-null float64
info_size             8361 non-null float64
info_lot_size         5931 non-null float64
info_year_built       8099 non-null float64
info_property_id      8763 non-null int64
info_county           8763 non-null object
info_parcel_number    8763 non-null int64
taxes_land            8632 non-null float64
taxes_improvements    8632 non-null float64
taxes_total           8632 non-null float64
taxes_taxes           8632 non-null object
school                8763 non-null int64
total_crime           84

`train_X_df` có `8763` dòng dữ liệu và `25` cột.

In [8]:
val_X_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3756 entries, 1540 to 4911
Data columns (total 25 columns):
address_street        3749 non-null object
address_locality      3756 non-null object
address_region        3756 non-null object
address_code          3756 non-null int64
date_sold             3756 non-null object
mortgage              3756 non-null float64
info_type             3756 non-null object
info_bedrooms         3412 non-null float64
info_bathrooms        3415 non-null float64
info_size             3585 non-null float64
info_lot_size         2553 non-null float64
info_year_built       3472 non-null float64
info_property_id      3756 non-null int64
info_county           3756 non-null object
info_parcel_number    3756 non-null int64
taxes_land            3689 non-null float64
taxes_improvements    3689 non-null float64
taxes_total           3689 non-null float64
taxes_taxes           3689 non-null object
school                3756 non-null int64
total_crime           365

`val_X_df` có `3756` dòng dữ liệu và `25` cột.

## 4.2. Quan sát và lựa chọn dữ liệu bằng kiến thức cá nhân trên tập train (`train_X_df` và `train_y_sr`)

### 4.2.1. Quan sát cột `address_street`

In [ ]:
# https://stackoverflow.com/questions/55105280/how-to-extract-apartment-from-address-in-pandas
pat = r'^(?P<number>\d+)?(?P<direction>.\w(?= ))?(?P<street>.+(?=\bapt|\bunit)|.+(?=#)|.+)(?P<apt_unit>(?:\bapt|\bunit|#|).+)?'
lower_address_street = train_X_df.address_street.str.lower()
tmp = lower_address_street.str.extract(pat)
tmp

In [ ]:
train_X_df.iloc[10]

In [ ]:
tmp["street"].value_counts()

In [ ]:
sum(tmp["street"].value_counts() > 1)

In [ ]:
len(tmp["street"].unique())

Theo kết quả tìm kiếm trên internet (https://cartographic.info/usa/street/show.php?p=california&t=Orange%20County) thì có `280` tên đường lớn tại `Orange County`. Vậy trong danh sách mà nhóm parse được có đến `6051` tên đường khác nhau, chắc bao gồm những con đường nhỏ. Việc dùng `street` cho huấn luyện mô hình sẽ không hiệu quả, vì khi đưa vào mô hình bằng phương pháp one-hot chắc chắn sẽ tạo ra số chiều rất lớn và với một số địa chỉ xuất hiện rất ít dễ dẫn đén overfitting.

In [ ]:
len(tmp[pd.notna(tmp["apt_unit"])])

Nhóm sẽ xóa cột `address_street` và thêm cột `apt_unit` được rút trích ở trên vào dữ liệu ở bước tiền xử lý dữ liệu, cột này chỉ có 2 giá trị (`True`/`False`)

### 4.2.2. Quan sát các cột `address_locality`, `address_region` và `address_code`

&#9889; Theo tìm hiểu của nhóm, `address_code` là tổng hợp từ `address_locality` và `address_region`

In [ ]:
unique_code = train_X_df.address_code.unique()
len(unique_code)

In [ ]:
for code in unique_code:
    tmp_df = train_X_df[train_X_df.address_code == code]
    print("Code: " + str(code))
    print("\n + Locality: " + str(tmp_df.address_locality.unique()))
    print("\n + Region: " + str(tmp_df.address_region.unique()))
    print("\n")

Theo nhóm tìm hiểu thì mã `code` (`address_code`) kia là mã bưu điện, mã này có thể dùng cho nhiều `locality` ở gần nhau, do đó ta kết quả như ở trên là hoàn toàn bình thường. Giờ nhóm quyết định bỏ 2 cột `address_locality` và `addres_region`, chỉ giữ lại cột `address_code` là đủ.

2. Xem thử cột `address_code` có bao nhiêu giá trị thiếu

In [ ]:
num_miss_code = train_X_df.shape[0] - len(train_X_df[pd.notna(train_X_df.address_code)])
print("The number of missing values (column 'address_code'): " + str(num_miss_code))

Good. Cột `address_code` không có giá trị thiếu.

### 4.2.3. Quan sát cột `date_sold`

Đây là cột dạng chuỗi có giá trị rời rạc không thứ tự (categorical)

1. Xem thử cột `date_sold` có bao nhiêu giá trị thiếu

In [14]:
num_miss_date_sold = train_X_df.shape[0] - len(train_X_df[pd.notna(train_X_df.date_sold)])
print("The number of missing values (column 'date_sold'): " + str(num_miss_date_sold))

The number of missing values (column 'date_sold'): 0


Good. Cột `date_sold` không có giá trị thiếu.

2. Rút trích thông tin từ cột `date_sold`

In [15]:
train_X_df.date_sold.unique()

array(['09/11/2019', '08/23/2019', '07/29/2019', '08/22/2019',
       '07/25/2019', '06/13/2019', '11/20/2019', '06/20/2019',
       '08/06/2019', '06/10/2019', '06/17/2019', '07/12/2019',
       '08/01/2019', '11/07/2019', '07/03/2019', '09/23/2019',
       '10/08/2019', '08/19/2019', '08/09/2019', '06/27/2019',
       '06/28/2019', '11/19/2019', '10/24/2019', '06/11/2019',
       '09/12/2019', '08/20/2019', '10/17/2019', '09/25/2019',
       '11/01/2019', '11/14/2019', '08/16/2019', '09/17/2019',
       '08/15/2019', '07/26/2019', '07/05/2019', '07/08/2019',
       '09/24/2019', '07/18/2019', '09/13/2019', '08/30/2019',
       '09/20/2019', '08/26/2019', '10/11/2019', '02/04/2019',
       '07/31/2019', '06/18/2019', '06/24/2019', '11/06/2019',
       '10/09/2019', '09/26/2019', '06/14/2019', '10/16/2019',
       '09/09/2019', '10/07/2019', '07/19/2019', '07/09/2019',
       '11/15/2019', '07/15/2019', '06/26/2019', '08/14/2019',
       '09/30/2019', '09/06/2019', '10/03/2019', '10/25

Cột `date_sold` được ghi dưới dạng `MM/DD/YYYY`. Giờ nhóm sẽ rút trích `Month` từ `date_sold`.

In [20]:
month = train_X_df.date_sold.str[:2].astype(int)
sorted(month.unique())

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

Sở dĩ không có tháng `12` là do thời gian lấy dữ liệu của nhóm là đầu tháng `12` nên trang web chưa cập nhật. Về sau ở bước `Xây dựng pipline tiền xử lý dữ liệu`, nhóm sẽ tiến hành xóa cột `date_sold`, thêm cột `Month` và muốn thử để nguyên `11` tháng huấn luyện tốt hơn hay chia ra thành `4` mùa tốt hơn bằng cách đặt thêm cờ hiệu `month_to_seasion`.

In [19]:
(month%12 + 3)//3

10747    4
12214    3
5916     3
489      3
12257    3
5010     3
2271     4
11574    3
6780     3
12148    3
6472     3
2560     3
6759     3
7948     3
263      4
7105     3
5013     3
658      4
1462     4
771      3
6403     3
2002     3
9628     3
4749     3
8001     4
3125     4
10326    3
12278    4
8851     3
3200     3
        ..
7599     4
1871     4
2046     4
7877     4
4851     3
5072     3
2163     4
6036     4
6921     3
6216     4
11085    4
537      3
9893     3
2897     3
7768     4
2222     3
10327    3
2599     3
705      3
3468     4
6744     3
5874     3
4373     4
7891     3
9225     4
4859     4
3264     4
9845     4
10799    4
2732     3
Name: date_sold, Length: 8763, dtype: int32

### 4.2.4. Quan sát cột `mortgage`

Đây là cột dạng số (numerical) - float64

`mortgage` là tiền cho thuê tính theo tháng của căn nhà khi chưa được bán, nó cũng gần gần để suy ra `sold_price` rồi.

In [ ]:
train_X_df.mortgage.describe()

Xem thử cột `mortgage` có bao nhiêu giá trị thiếu

In [ ]:
num_miss_mortgage = train_X_df.shape[0] - len(train_X_df[pd.notna(train_X_df.mortgage)])
print("The number of missing values (column 'mortgage'): " + str(num_miss_mortgage))

Good. Cột `mortgage` không có giá trị thiếu.

### 4.2.5. Quan sát cột `info_type`

Đây là cột dạng chuỗi có giá trị rời rạc không thứ tự (categorical)

1. Quan sát các giá trị xuất hiện trong cột và số lượng tương ứng

In [ ]:
train_X_df.info_type.value_counts()

Có một số giá trị xuất hiện rất ít, nhóm thử dùng thêm tham số `num_top_type` để chọn ra các `info_type` xuất hiện nhiều lần; còn các giá trị nằm ngoài `num_top_type` nhóm sẽ cho là giá trị `Others`. (như `BT03-TienXuLy_ChongOverfit`)

2. Xem thử cột `address_type` có bao nhiêu giá trị thiếu

In [ ]:
num_miss_type = train_X_df.shape[0] - len(train_X_df[pd.notna(train_X_df.info_type)])
print("The number of missing values (column 'info_type'): " + str(num_miss_type))

Good. Cột `info_type` không có giá trị thiếu.

### 4.2.5. Quan sát cột `info_bedrooms` và cột `info_bathrooms`

Đây là cột dạng số (numerical) - float64

1. Xem thử cột `info_bedrooms` có bao nhiêu giá trị thiếu và chọn cách để lắp đầy các giá trị thiếu đó.

In [ ]:
num_miss_bedrooms = train_X_df.shape[0] - len(train_X_df[pd.notnull(train_X_df.info_bedrooms)])
print("The number of missing values (column 'info_bedrooms'): " + str(num_miss_bedrooms))

Cột `info_bedrooms` có `812` giá trị bị thiếu.

In [ ]:
train_X_df.info_bedrooms.value_counts()

In [ ]:
train_X_df.info_bedrooms.describe()

In [ ]:
train_X_df.info_bedrooms.value_counts().sort_index().plot(kind='bar')

Do là `bedrooms` và các giá trị của nó là số nguyên, nên nhóm dùng `most` để lắp đầy các giá trị thiếu cho cột `info_bedrooms`

2. Xem thử cột `info_bathrooms` có bao nhiêu giá trị thiếu và chọn cách để lắp đầy các giá trị thiếu đó.

In [ ]:
num_miss_bathrooms = train_X_df.shape[0] - len(train_X_df[pd.notnull(train_X_df.info_bathrooms)])
print("The number of missing values (column 'info_bathrooms'): " + str(num_miss_bathrooms))

Cột `info_bathrooms` có `807` giá trị bị thiếu.

In [ ]:
train_X_df.info_bathrooms.value_counts()

In [ ]:
train_X_df.info_bathrooms.describe()

In [ ]:
train_X_df.info_bathrooms.value_counts().sort_index().plot(kind='bar')

Tương tự nhóm cùng sẽ dùng `most` để lắp đầy các giá trị thiếu cho cột `info_bathrooms`

### 4.2.6. Quan sát cột `info_size` và cột `info_lot_size`

Đây là cột dạng số (numerical) - float64

1. Xem thử cột `info_size` có bao nhiêu giá trị thiếu và chọn cách để lắp đầy các giá trị thiếu đó.

In [ ]:
num_miss_size= train_X_df.shape[0] - len(train_X_df[pd.notnull(train_X_df.info_size)])
print("The number of missing values (column 'info_size'): " + str(num_miss_size))

Cột `info_size` có `402` giá trị bị thiếu.

In [ ]:
train_X_df.info_size.value_counts()

In [ ]:
train_X_df.info_size.describe()

In [ ]:
len(train_X_df.info_size.unique())

Điền các giá trị thiếu bằng `mean`.

2. Xem thử cột `info_lot_size` có bao nhiêu giá trị thiếu và chọn cách để lắp đầy các giá trị thiếu đó.

In [ ]:
num_miss_lot_size= train_X_df.shape[0] - len(train_X_df[pd.notnull(train_X_df.info_lot_size)])
print("The number of missing values (column 'info_size'): " + str(num_miss_lot_size))

Cột `info_size` có `2832` giá trị bị thiếu.

In [ ]:
train_X_df.info_lot_size.value_counts()

In [ ]:
train_X_df.info_lot_size.describe()

In [ ]:
len(train_X_df.info_lot_size.unique())

Điền các giá trị thiếu bằng `mean`.

### 4.2.7. Quan sát cột `info_year_built`

Đây là cột dạng số (numerical) - float64

Xem thử cột `info_year_built` có bao nhiêu giá trị thiếu

In [ ]:
num_miss_year_built = train_X_df.shape[0] - len(train_X_df[pd.notnull(train_X_df.info_year_built)])
print("The number of missing values (column 'info_size'): " + str(num_miss_year_built))

Cột `info_year_built` có `664` giá trị bị thiếu.

In [ ]:
train_X_df.info_year_built.value_counts()

In [ ]:
train_X_df.info_year_built.describe()

In [ ]:
len(train_X_df.info_year_built.unique())

### 4.2.8. Quan sát các cột `info_property_id`, `info_county` và `info_parcel_number`

- `info_property_id` là cột dạng số (numerical) - int64
- `info_county` là cột dạng chuỗi có giá trị rời rạc không thứ tự (categorical)
- `info_parcel_number` là cột dạng số (numerical) - int64

Bỏ đi các cột này, vì 2 cột `info_property_id` và `info_parcel_number` ứng với mỗi dòng dữ liệu là hoàn toàn khác nhau; còn cột `info_county` thì có giá trị là `Orange County` (do dữ liệu thu thập của Quận Cam).

### 4.2.9. Quan sát các cột `taxes_land`, `taxes_improvements`, `taxes_total` và `taxes_taxes`

Các cột này đều là cột dạng số (numerical) - float64

In [ ]:
num_miss_taxes = train_X_df.shape[0] - len(train_X_df[pd.notnull(train_X_df.taxes_land)])
print("The number of missing values (column 'taxes'): " + str(num_miss_taxes))

Có `131` giá trị thiếu cho mỗi cột `taxes_land`, `taxes_improvements`, `taxes_total` và `taxes_taxes`

Search cùng `address_code` thì lắp đầy các giá trị thiếu bằng `mean` của các `taxes` trong cùng `address_code`

Giữ nguyên các cột `taxes_land`, `taxes_improvements` và bỏ cột `taxes_total`. Giải thích thêm ý nghĩa của các cột này:
- Cột `taxes_land` là tiền cải thuế cải tạo đất mà mình phải đóng cho chính phủ.
- Cột `taxes_improvements` là tiền mà mình phải trả do sử dụng các tiện ích của chính phủ (cắt cỏ, tiền rác, củng cố hàng rào,...)
- Cột `taxes_total` là tổng của 2 cột trên.

Đối với cột `taxes_taxes` thì nhóm sẽ rút trích số tiền phải trả trong cột này ra.

In [ ]:
train_X_df.taxes_taxes

In [ ]:
train_X_df.taxes_taxes.str.extract(r'(\d+)', expand=False)

### 4.2.10. Quan sát cột `school` và cột `foreclosures`

Các cột này đều là cột dạng số (numerical) - int64

Hai cột này không có giá trị thiếu, nên được giữ nguyên để huấn luyện.

In [ ]:
train_X_df.school.value_counts().sort_index().plot(kind='bar')

In [ ]:
train_X_df.foreclosures.value_counts().sort_index().plot(kind='bar')

### 4.2.11. Quan sát cột `total_crime`, `violent_crime` và `property_crime`

Các cột này đều là cột dạng số (numerical) - float64

Giải thích về ý nghĩa của các cột này. Giá trị mà các cột `total_crime`, `violent_crime` và `property_crime` biểu diễn được tính theo % so với tỉ lệ tội phạm quốc gia. Ví dụ `total_crime = 79` thì ở đây tổng số tội phạm là `79%` so với toàn nước (`100%`).

In [ ]:
num_miss_crime = train_X_df.shape[0] - len(train_X_df[pd.notnull(train_X_df.total_crime)])
print("The number of missing values (column 'taxes'): " + str(num_miss_crime))

Có `265` giá trị thiếu cho mỗi cột `total_crime`, `violent_crime` và `property_crime`

In [ ]:
train_X_df.total_crime.value_counts()

Search cùng `address_code` thì lắp đầy các giá trị thiếu bằng `median` của các `crime` trong cùng `address_code`

### 4.2.12. Quan sát `correct input` (`sold_price` của ` train_y_sr`)

Đây là cột dạng số (numerical) - float64

In [ ]:
train_y_sr.describe()

In [ ]:
train_y_sr.sort_values().sort_index().plot(style='.')

## 4.3. Thiết kế pipeline tiền xử lý dữ liệu

1. Cột `address_street`: 
    - Ở đây có 4 thông tin được rút trích (xem  phần `4.2.1. Quan sát cột address_street` ở trên):
         + `number`: số nhà, thông tin này không cần thiết cho việc huấn luyện.
         + `direction`: hướng (đông, tây, nam, bắc) của đường, thông tin này không cần thiết cho việc huấn luyện.
         + `street`: tên đường, theo kết quả parse thì thu được đến 6051 tên đường lớn/nhỏ khác nhau. Nếu đưa tên đường này vào huấn luyện thì có 3 vấn đề: thứ nhất là nếu dùng one-hot để chuyển sang dạng số thì sẽ làm tăng số chiều dữ liệu lên rất nhiều (thêm đến trên 6000 thuộc tính), thứ hai do có khoảng 1500 tên đường chỉ xuất hiện duy nhất một lần trong `train_X_df`, còn các tên đường khác thì cũng có tần số xuất hiện khá thấp (cao nhất chỉ có 16 lần) nên nếu thêm vào huấn luyện có thể dẫn đến overfitting, thứ 3 là nếu như ở tập `validation` hay tập `test` mà có tên đường không nằm trong tập `train` thì cũng khó khăn để giải quyết. Nên nhóm quyết định không lấy thông tin `street` để huấn luyện.
         + `apt_unit`: cho biết căn nhà có thuộc chung cư nào hay không, thông tin này nhóm nghĩ là cần thiết cho việc huấn luyện vì nghĩ giá nhà riêng sẽ có phần nào khác với các căn nhà thuộc các khu chung cư. Vì vậy, sẽ thêm cột `is_apt` vào để huấn luyện, cột này chỉ có 2 giá trị `0`/`1` (nếu thuộc chung cư thì `1`, ngược lại thì `0`).
    - Tóm lại, các việc cần làm:
        + Rút trích ra cột `is_apt` (như mô tả ở trên) và thêm vào dữ liệu.
        + Xóa cột `address_street`.
        + Lắp đầy giá trị thiếu bằng KNN (n=5).
        + Đây là dữ liệu số, để nguyên vậy vào huấn luyện.
         
2. Xóa 2 cột `address_locality` và cột `address_region`, giữ lại cột `address_code` (lý do xem ở phần `4.2.2. Quan sát các cột address_locality, address_region và address_code` ở trên). Ở tập `train_X_df` thì không thấy dữ liệu thiếu, nếu thiếu dùng KNN (n=5) để lấp đầy. Đưa vào tập train huấn luyện thì xem tại link (https://towardsdatascience.com/transforming-categorical-information-into-usable-measures-in-a-machine-learning-model-e2910bbb3fc7).
3. Cột `date_sold`: rút trích `month` ra (xem ở phần `4.2.3. Quan sát cột date_sold`), xóa cột `date_sold` và `year_sol`, thêm vào cột `month_sold`. Ở tập `train_X_df` thì không thấy dữ liệu thiếu, nếu thiếu dùng KNN(n=5) để lấp đầy. Bỏ vào huấn luyện dùng one-hot, nhớ là thử để nguyên `11` tháng huấn luyện tốt hơn hay chia ra thành `4` mùa tốt hơn bằng cách đặt thêm cờ hiệu `month_to_season`.
4. Cột `mortgage`: để nguyên. Ở tập `train_X_df` thì không thấy dữ liệu thiếu, nếu thiếu dùng KNN (n=5) để lấp đầy. Dạng số, cứ bỏ thẳng vào huấn luyện.
5. Cột `info_type`: Có một số giá trị xuất hiện rất ít, nhóm thử dùng thêm tham số `num_top_types` để chọn ra các `info_type` xuất hiện nhiều lần; còn các giá trị nằm ngoài `num_top_types` nhóm sẽ cho là giá trị `Others`. (như `BT03-TienXuLy_ChongOverfit`). Ở tập `train_X_df` thì không thấy dữ liệu thiếu, nếu thiếu dùng KNN (n=5) để lấp đầy. Đưa vào huấn luyện thì dùng one-hot.
6. Hai cột `info_bedrooms` và `info_bathrooms`: lắp đầy giá trị thiều bằng `most`. Kiểu số, cứ bỏ thẳng vào huấn luyện.
7. Hai cột `info_size` và `info_lot_size`: dùng KNN (n=5) để lắp đầy giá trị thiếu. Kiểu số, cứ bỏ thẳng vào huấn luyện.
8. Xóa 3 cột` info_property_id`, `info_county` và `info_parcel_number` (lý do xem ở phần `4.2.9. Quan sát các cột info_property_id, info_county và info_parcel_number`).
9. Các cột `taxes_land`, `taxes_improvements`, `taxes_total` và `taxes_taxes`: đầu tiên rút trích `taxes` cho cột `taxes_taxes`, thay bằng dữ liệu tìm được (xem ở phần `4.2.10. Quan sát các cột taxes_land, taxes_improvements và taxes_taxes`), nhớ xóa cột `taxes_total`. Điền các giá trị thiếu bằng `mean`. Kiểu số, cứ bỏ thẳng vào huấn luyện.
10. Giữ nguyên cột `school` và cột `foreclosures`. Nếu có giá trị thiếu thì điền bằng `median`. Kiểu số, cứ bỏ thẳng vào huấn luyện.
11. Các `total_crime`, `violent_crime` và `property_crime`: lắp đầy giá trị thiếu bằng `median`. Kiểu số, cứ bỏ thẳng vào huấn luyện.

- Chuẩn hóa theo cách của thầy trong `BT03`.
- Dùng mô hình `Linear Regression` để huấn luyện mô mình thử xem.

In [9]:
import sklearn
sklearn.__version__

'0.22'

In [10]:
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.impute import SimpleImputer, KNNImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LinearRegression

In [11]:
# Class AddressProcessor dùng để xử lý các cột address_street, address_locality và address_region.
# Được kế thừa từ 2 class của Sklearn là BaseEstimator và TransformerMixin. 
# Việc kế thừa này giúp class của ta tự động có các phương thức như set_params, get_params, fit_transform.
class AddressProcessor(BaseEstimator, TransformerMixin):
    def __init__(self):
        pass

    def fit(self, X_df, y=None):
        return self

    def transform(self, X_df, y=None):
        # Extract các thông tin (số nhà, hướng của đường, tên đường, tên chung cư) từ address_street.
        pat = r'^(?P<number>\d+)?(?P<direction>.\w(?= ))?(?P<street>.+(?=\bapt|\bunit)|.+(?=#)|.+)(?P<apt_unit>(?:\bapt|\bunit|#|).+)?'
        lower_address_street = X_df.address_street.str.lower()
        street_parsed = lower_address_street.str.extract(pat)
        
        # Xóa các cột dùng nữa.
        transformed_df = X_df.drop(["address_street", "address_locality", "address_region"], axis=1)
        
        # Thêm cột is_apt vào DataFrame (1 là thuộc chung cư, 0 là không không thuộc chung cư).
        transformed_df["is_apt"] = street_parsed["apt_unit"].notnull().astype(int)
        return transformed_df

In [28]:
# TEST AddressProcessor
ap = AddressProcessor()

ap.transform(train_X_df).is_apt.value_counts()

0    7285
1    1478
Name: is_apt, dtype: int64

In [33]:
# Class DateSoldProcessor dùng để xử lý các cột date_sold và year_sold.
# Được kế thừa từ 2 class của Sklearn là BaseEstimator và TransformerMixin. 
# Việc kế thừa này giúp class của ta tự động có các phương thức như set_params, get_params, fit_transform.

class DateSoldProcessor(BaseEstimator, TransformerMixin):
    def __init__(self, month_to_season = False):
        self.month_to_season = month_to_season

    def fit(self, X_df, y=None):
        return self

    def transform(self, X_df, y=None):
        # Rút trích month từ date_sold.
        month = X_df.date_sold.str[:2].astype(int)
        
        # Xóa các cột không dùng nữa.
        transformed_df = X_df.drop(["date_sold", "year_sold"], axis=1)
        
        if (self.month_to_season == False):
            # Thêm cột month vào DataFrame (để tên là time_sold cho tiện truy cập).
            transformed_df["time_sold"] = month
        else:
            # 1 là Đông (tháng 12, 1, 2)
            # 2 là Xuân (tháng 3, 4, 5)
            # 3 là Hạ (tháng 6, 7, 8)
            # 4 là Thu (tháng 9, 10, 11)
            season = ((month.astype(int)%12 + 3)//3).astype(str)
            # Thêm cột season_sold vào DataFrame (để tên là time_sold cho tiện truy cập).
            transformed_df["time_sold"] = season
        return transformed_df

In [34]:
# TEST DateSoldProcessor
dsp = DateSoldProcessor()

dsp.transform(train_X_df).time_sold.value_counts()

7     1719
8     1707
9     1593
10    1356
6     1283
11    1068
5       12
4        9
3        7
2        5
1        4
Name: time_sold, dtype: int64

In [36]:
# TEST DateSoldProcessor
dsp = DateSoldProcessor(True)

dsp.transform(train_X_df).time_sold.value_counts()

3    4709
4    4017
2      28
1       9
Name: time_sold, dtype: int64

In [29]:
# Class InfoTypeProcessor dùng để xử lý cột info_type.
# Được kế thừa từ 2 class của Sklearn là BaseEstimator và TransformerMixin. 
# Việc kế thừa này giúp class của ta tự động có các phương thức như set_params, get_params, fit_transform.

class InfoTypeProcessor(BaseEstimator, TransformerMixin):
    def __init__(self, num_top_types=1):
        self.num_top_types = num_top_types

    def fit(self, X_df, y=None):
        info_type_col = X_df.info_type
        
        # Đếm các dạng type trong cột info_type
        self.type_counts_ = info_type_col.value_counts()
        types = list(self.type_counts_.index)
        
        # Lấy ra top_type theo tham số num_top_types được truyền vào
        self.top_types_ = types[:max(1, min(self.num_top_types, len(types)))]
        return self

    def transform(self, X_df, y=None):
        # Các cột không phải top_type được mặc định là "Others"
        type_sr = X_df.info_type.str.extract(f'^({"|".join(self.top_types_)})$', expand=False).fillna("Others")
        
        # Xóa đi cột info_type vì không dùng nữa.
        transformed_df = X_df.drop(["info_type"], axis=1)
        
        # Thêm cột type vào DataFrame.
        transformed_df["type"] = type_sr
        return transformed_df

In [30]:
# TEST InfoTypeProcessor
itp = InfoTypeProcessor(num_top_types=5)

itp.fit(train_X_df).transform(train_X_df).type.value_counts()

Single Family Residence    5709
Condominium                2399
MISCELLANEOUS               326
Multi-Family Dwellings      230
Contact Agent                70
Others                       29
Name: type, dtype: int64

In [31]:
# Class TaxesProcessor dùng để xử lý các cột taxes_total và taxes_taxes.
# Được kế thừa từ 2 class của Sklearn là BaseEstimator và TransformerMixin. 
# Việc kế thừa này giúp class của ta tự động có các phương thức như set_params, get_params, fit_transform.

class TaxesProcessor(BaseEstimator, TransformerMixin):
    def __init__(self):
        pass

    def fit(self, X_df, y=None):
        return self

    def transform(self, X_df, y=None):
        # Extract ra tiền thuế phải trả.
        taxes = X_df.taxes_taxes.str.extract(r'(\d+)', expand=False)
        
        # Xóa đi các cột không cần thiết.
        transformed_df = X_df.drop(["taxes_total", "taxes_taxes"], axis=1)
        
        # Thêm cột taxes_taxes với giá trị mới extract được vào.
        transformed_df["taxes_taxes"] = taxes
        return transformed_df

In [32]:
# TEST InfoTypeProcessor
tp = TaxesProcessor()

tp.fit(train_X_df).transform(train_X_df).taxes_taxes.value_counts()

10555     16
7259       8
5026       7
1014       6
4127       6
6076       6
4973       6
4875       6
3759       5
2669       5
3688       5
1310       5
2545       5
1152       5
3011       5
5753       5
2749       5
6029       5
7190       5
6540       5
4986       5
3634       5
3606       5
6333       4
1244       4
5901       4
4301       4
5076       4
4356       4
5695       4
          ..
2274       1
16588      1
7384       1
12839      1
11296      1
995        1
9073       1
4960       1
5301       1
28593      1
3051       1
26568      1
12497      1
19356      1
2884       1
12434      1
16118      1
162919     1
4755       1
12914      1
5489       1
8358       1
7377       1
5034       1
1977       1
7792       1
13771      1
4242       1
16113      1
35333      1
Name: taxes_taxes, Length: 6348, dtype: int64

In [87]:
# Class ColumnDropper dùng để xóa các cột info_property_id, info_county, info_parcel_number và mortgage.
# Được kế thừa từ 2 class của Sklearn là BaseEstimator và TransformerMixin. 
# Việc kế thừa này giúp class của ta tự động có các phương thức như set_params, get_params, fit_transform.
class ColumnDropper(BaseEstimator, TransformerMixin):
    def __init__(self):
        pass

    def fit(self, X_df, y=None):
        return self

    def transform(self, X_df, y=None):
        transform_df = X_df.drop(["info_property_id", "info_county", "info_parcel_number", "mortgage"], axis=1)
        #transform_df = X_df.drop(["info_property_id", "info_county", "info_parcel_number"], axis=1)
        transform_df["address_code"] = transform_df["address_code"].astype("str")
        return transform_df

In [84]:
# TEST InfoTypeProcessor
cd = ColumnDropper()

cd.fit(train_X_df).transform(train_X_df).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8763 entries, 10747 to 2732
Data columns (total 22 columns):
address_street        8736 non-null object
address_locality      8763 non-null object
address_region        8763 non-null object
address_code          8763 non-null object
date_sold             8763 non-null object
mortgage              8763 non-null float64
info_type             8763 non-null object
info_bedrooms         7951 non-null float64
info_bathrooms        7956 non-null float64
info_size             8361 non-null float64
info_lot_size         5931 non-null float64
info_year_built       8099 non-null float64
taxes_land            8632 non-null float64
taxes_improvements    8632 non-null float64
taxes_total           8632 non-null float64
taxes_taxes           8632 non-null object
school                8763 non-null int64
total_crime           8498 non-null float64
violent_crime         8498 non-null float64
property_crime        8498 non-null float64
foreclosures      

In [ ]:
"""
class ColumnTransformerWithBaseColumn(BaseEstimator, TransformerMixin):
    def __init__(self, colTransformer, baseCol):
        self.colTransformer = colTransformer
        self.baseCol = baseCol

    def fit(self, X_df, y=None):
        self.train_X_df = X_df
        return self

    def transform(self, X_df, y=None):
        colTransformer.fit()
        transformed_df = X_df.copy()
        transformed_df["type"] = type_sr
        return transformed_df
"""

In [88]:
knn_cols = ["info_size", "info_lot_size"]
most_frequent_cols = ["info_bedrooms", "info_bathrooms"]
mean_cols = ["taxes_land", "taxes_improvements", "taxes_taxes"]
median_cols = ["school", "foreclosures", "total_crime", "violent_crime", "property_crime"]

code_pipeline = Pipeline([("most_frequent", SimpleImputer(strategy="most_frequent")),
                          ("onehot", OneHotEncoder())])

type_pipeline = Pipeline([("most_frequent", SimpleImputer(strategy="most_frequent")),
                          ("onehot", OneHotEncoder())])

time_pipeline = Pipeline([("most_frequent", SimpleImputer(strategy="most_frequent")),
                           ("onehot", OneHotEncoder())])


column_transformer = ColumnTransformer([("knn", KNNImputer(), knn_cols),
                                        ("most_frequent", SimpleImputer(strategy="most_frequent"), most_frequent_cols),
                                        ("mean", SimpleImputer(strategy="mean"), mean_cols),
                                        ("median", SimpleImputer(strategy="median"), median_cols),
                                        ("code_col", code_pipeline, ["address_code"]),
                                        ("type_col", type_pipeline, ["type"]),
                                        ("time_col", time_pipeline, ["time_sold"])])


preprocess_pipeline = Pipeline([("address", AddressProcessor()),
                                ("date_sold", DateSoldProcessor()),
                                ("info_type", InfoTypeProcessor()),
                                ("col_dropper", ColumnDropper()),
                                ("taxes", TaxesProcessor()),
                                ("column_transformer", column_transformer)])


preprocess_pipeline.fit_transform(train_X_df)

<8763x112 sparse matrix of type '<class 'numpy.float64'>'
	with 130533 stored elements in Compressed Sparse Row format>

In [89]:
full_pipeline = Pipeline([("preprocess", preprocess_pipeline),
                          ("std_scaler", StandardScaler(with_mean=False)),
                          ("regression", LinearRegression())])

full_pipeline.fit(train_X_df, train_y_sr)
full_pipeline.score(val_X_df, val_y_sr)

0.3676110686066309

In [68]:
from sklearn.neighbors import KNeighborsRegressor

N = [1, 3, 5, 7, 9, 11]
for n in N:
    full_pipeline = Pipeline([("preprocess", preprocess_pipeline),
                              ("std_scaler", StandardScaler(with_mean=False)),
                              ("regression", KNeighborsRegressor(n_neighbors=n))])

    full_pipeline.fit(train_X_df, train_y_sr)
    accuracy = full_pipeline.score(val_X_df, val_y_sr)
    print("With n = " + str(n) + ": " + str( accuracy))

With n = 1: 0.7219044459616868
With n = 3: 0.7260192789749167
With n = 5: 0.6924391144956822
With n = 7: 0.5977965056021486
With n = 9: 0.5426789336279418
With n = 11: 0.4905314288717677


In [69]:
from sklearn.linear_model import ElasticNet
full_pipeline = Pipeline([("preprocess", preprocess_pipeline),
                          ("std_scaler", StandardScaler(with_mean=False)),
                          ("regression", ElasticNet())])

full_pipeline.fit(train_X_df, train_y_sr)
full_pipeline.score(val_X_df, val_y_sr)

0.20461060052249203

In [70]:
from sklearn.tree import DecisionTreeRegressor

for n in range(50):
    full_pipeline = Pipeline([("preprocess", preprocess_pipeline),
                              ("std_scaler", StandardScaler(with_mean=False)),
                              ("regression", DecisionTreeRegressor(random_state=n))])

    full_pipeline.fit(train_X_df, train_y_sr)
    accuracy = full_pipeline.score(val_X_df, val_y_sr)
    print("With random_state = " + str(n) + ": " + str( accuracy))

With random_state = 0: 0.7925063653485082
With random_state = 1: 0.6659847554537413
With random_state = 2: 0.7839739401123859
With random_state = 3: 0.6609784165823651
With random_state = 4: 0.7839877423767306
With random_state = 5: 0.6620189136073906
With random_state = 6: 0.6515646438145635
With random_state = 7: 0.6627584522469738
With random_state = 8: 0.7872266236413078
With random_state = 9: 0.7908534991879248
With random_state = 10: 0.7838717840200333
With random_state = 11: 0.7847374034637077
With random_state = 12: 0.7878457517705085
With random_state = 13: 0.6640095679046969
With random_state = 14: 0.6539126524463794
With random_state = 15: 0.6562580072686481
With random_state = 16: 0.7808635129877541
With random_state = 17: 0.6601832600773501
With random_state = 18: 0.6600117004329101
With random_state = 19: 0.6594240874107244
With random_state = 20: 0.7936147602626978
With random_state = 21: 0.7903337310328472
With random_state = 22: 0.7803918817111278
With random_state = 2

In [71]:
from sklearn.tree import DecisionTreeRegressor

for n in range(49,101):
    full_pipeline = Pipeline([("preprocess", preprocess_pipeline),
                              ("std_scaler", StandardScaler(with_mean=False)),
                              ("regression", DecisionTreeRegressor(random_state=n))])

    full_pipeline.fit(train_X_df, train_y_sr)
    accuracy = full_pipeline.score(val_X_df, val_y_sr)
    print("With random_state = " + str(n) + ": " + str( accuracy))

With random_state = 49: 0.7846271618653804
With random_state = 50: 0.657705717768456
With random_state = 51: 0.7903701063816483
With random_state = 52: 0.789463193365913
With random_state = 53: 0.7848317982743029
With random_state = 54: 0.7821225563607986
With random_state = 55: 0.780751800274875
With random_state = 56: 0.7865036031389248
With random_state = 57: 0.6680343865816866
With random_state = 58: 0.7885686877880836
With random_state = 59: 0.7828092171647087
With random_state = 60: 0.6570193652062462
With random_state = 61: 0.7812353669146882
With random_state = 62: 0.7916158440034105
With random_state = 63: 0.7864809773376831
With random_state = 64: 0.7824400581118248
With random_state = 65: 0.7883876450025604
With random_state = 66: 0.7781623366322544
With random_state = 67: 0.7864453557323444
With random_state = 68: 0.7811652603708673
With random_state = 69: 0.7793856210004718
With random_state = 70: 0.6679552342903046
With random_state = 71: 0.784203557141179
With random_sta

In [75]:
from sklearn.ensemble import GradientBoostingRegressor

for n in range(0,51):
    full_pipeline = Pipeline([("preprocess", preprocess_pipeline),
                              ("std_scaler", StandardScaler(with_mean=False)),
                              ("regression", GradientBoostingRegressor(random_state=n))])

    full_pipeline.fit(train_X_df, train_y_sr)
    accuracy = full_pipeline.score(val_X_df, val_y_sr)
    print("With random_state = " + str(n) + ": " + str( accuracy))

With random_state = 0: 0.7999780262108753
With random_state = 1: 0.7867585720678776
With random_state = 2: 0.8350462151858227
With random_state = 3: 0.800456280158938
With random_state = 4: 0.7857022131635122
With random_state = 5: 0.8161527277201609
With random_state = 6: 0.7975735193356739
With random_state = 7: 0.7995885075806872
With random_state = 8: 0.8091569923327382
With random_state = 9: 0.8270965165055384
With random_state = 10: 0.8088404087140052
With random_state = 11: 0.8224852445178769
With random_state = 12: 0.8116496251875935
With random_state = 13: 0.8006188520567398
With random_state = 14: 0.8154819436159858
With random_state = 15: 0.8120454536422175
With random_state = 16: 0.8076591079462648
With random_state = 17: 0.8286823148798782
With random_state = 18: 0.7925045577657999
With random_state = 19: 0.8116078600836041
With random_state = 20: 0.7977932009328877
With random_state = 21: 0.8074731858629904
With random_state = 22: 0.7858813281110458
With random_state = 23

https://www.kaggle.com/junkal/selecting-the-best-regression-model

In [76]:
from sklearn.ensemble import RandomForestRegressor
for n in range(10, 21):
    full_pipeline = Pipeline([("preprocess", preprocess_pipeline),
                              ("std_scaler", StandardScaler(with_mean=False)),
                              ("regression", RandomForestRegressor(n_estimators=n))])

    full_pipeline.fit(train_X_df, train_y_sr)
    accuracy = full_pipeline.score(val_X_df, val_y_sr)
    print("With n_trees = " + str(n) + ": " + str( accuracy))

With n_trees = 10: 0.7680782711703396
With n_trees = 11: 0.7941500164790997
With n_trees = 12: 0.7428607908578331
With n_trees = 13: 0.8151927248615991
With n_trees = 14: 0.7687711487859537
With n_trees = 15: 0.8251480930437328
With n_trees = 16: 0.7934635073460735
With n_trees = 17: 0.7571777419993032
With n_trees = 18: 0.7928195966532341
With n_trees = 19: 0.7839380825664696
With n_trees = 20: 0.7516249047511813


In [78]:
from sklearn.ensemble import RandomForestRegressor
for n in range(21, 31):
    full_pipeline = Pipeline([("preprocess", preprocess_pipeline),
                              ("std_scaler", StandardScaler(with_mean=False)),
                              ("regression", RandomForestRegressor(n_estimators=n))])

    full_pipeline.fit(train_X_df, train_y_sr)
    accuracy = full_pipeline.score(val_X_df, val_y_sr)
    print("With n_trees = " + str(n) + ": " + str( accuracy))

With n_trees = 21: 0.7994778835506996
With n_trees = 22: 0.7826073302057023
With n_trees = 23: 0.7818011289822785
With n_trees = 24: 0.7853851100679016
With n_trees = 25: 0.776069950490403
With n_trees = 26: 0.8279547184418456
With n_trees = 27: 0.8068720668003042
With n_trees = 28: 0.8377148658342689
With n_trees = 29: 0.8251455026740729
With n_trees = 30: 0.7927289643855973


In [79]:
from sklearn.ensemble import RandomForestRegressor
for n in range(21):
    full_pipeline = Pipeline([("preprocess", preprocess_pipeline),
                              ("std_scaler", StandardScaler(with_mean=False)),
                              ("regression", RandomForestRegressor(random_state=n))])

    full_pipeline.fit(train_X_df, train_y_sr)
    accuracy = full_pipeline.score(val_X_df, val_y_sr)
    print("With random_state = " + str(n) + ": " + str( accuracy))

With random_state = 0: 0.7977509742093444
With random_state = 1: 0.8093499910266685
With random_state = 2: 0.8051308581115851
With random_state = 3: 0.8063576555764645
With random_state = 4: 0.7819483132699812
With random_state = 5: 0.8093965655598084
With random_state = 6: 0.8116561133660033
With random_state = 7: 0.8211473233633331
With random_state = 8: 0.7999900312565849
With random_state = 9: 0.8091801674703056
With random_state = 10: 0.7996988971647787
With random_state = 11: 0.816591975409707
With random_state = 12: 0.7972444990900895
With random_state = 13: 0.7885888275601943
With random_state = 14: 0.8078394361892346
With random_state = 15: 0.8036448802564152
With random_state = 16: 0.8248270254819615
With random_state = 17: 0.8057498429239374
With random_state = 18: 0.7976152645704063
With random_state = 19: 0.7913858077929808
With random_state = 20: 0.8132527716351619
